src/utils/upload/processUploadV1.js

```diff
@@ -0,0 +1,272 @@
+    );
+
+    dispatch(
+      updateSampleFile(
+        sampleUuid,
+        fileName,
+        { upload: { status: UploadStatus.UPLOADING, amplifyPromise: uploadPromise } },
+      ),
+    );
+
+    await uploadPromise;
+  } catch (e) {
+    // File size and file object should be available so we can reupload
+    dispatch(
+      updateSampleFile(
+        sampleUuid,
+        fileName,
+        {
+          ...filePropertiesToUpdate,
+          upload: { status: UploadStatus.UPLOAD_ERROR, amplifyPromise: null },
+        },
+      ),
+    );
+
+    return;
+  }
+
+  dispatch(
+    updateSampleFile(
+      sampleUuid,
+      fileName,
+      {
+        ...filePropertiesToUpdate,
+        upload: {
+          status: UploadStatus.UPLOADED,
+          progress: 100,
+          amplifyPromise: null,
+        },
+      },
+    ),
+  );
+};
+
+const renameFileIfNeeded = (fileName, type) => {
+  // rename files to include .gz
+  const uncompressed = !['application/gzip', 'application/x-gzip'].includes(type) && !fileName.endsWith('.gz');
+  let newFileName = uncompressed ? `${fileName}.gz` : fileName;
+
+  // We rename genes.tsv files to features.tsv (for a single entry)
+  newFileName = newFileName.replace('genes', 'features');
+
+  return newFileName;
+};
+
+const uploadSingleFile = (newFile, activeProjectUuid, sampleUuid, dispatch) => {
+  const metadata = metadataForFile(newFile);
+
+  const newFileName = renameFileIfNeeded(newFile.fileObject.name, newFile.fileObject.type);
+
+  compressAndUploadSingleFile(
+    activeProjectUuid, sampleUuid, newFileName, newFile, dispatch, metadata,
+  );
+
+  return [newFileName, { ...newFile, name: newFileName }];
+};
+
+const compressAndUpload = (sample, activeProjectUuid, dispatch) => Object.fromEntries(
+  Object.values(sample.files)
+    .map((file) => uploadSingleFile(file, activeProjectUuid, sample.uuid, dispatch)),
+);
+
+const processUpload = async (filesList, sampleType, samples, activeProjectUuid, dispatch) => {
+  const samplesMap = filesList.reduce((acc, file) => {
+    const pathToArray = file.name.trim().replace(/[\s]{2,}/ig, ' ').split('/');
+
+    const sampleName = pathToArray[0];
+    const fileName = _.last(pathToArray);
+
+    // Update the file name so that instead of being saved as
+    // e.g. WT13/matrix.tsv.gz, we save it as matrix.tsv.gz
+    file.name = fileName;
+
+    const sampleUuid = Object.values(samples).filter(
+      (s) => s.name === sampleName
+        && s.projectUuid === activeProjectUuid,
+    )[0]?.uuid;
+
+    return {
+      ...acc,
+      [sampleName]: {
+        ...acc[sampleName],
+        uuid: sampleUuid,
+        files: {
+          ...acc[sampleName]?.files,
+          [fileName]: file,
+        },
+      },
+    };
+  }, {});
+
+  Object.entries(samplesMap).forEach(async ([name, sample]) => {
+    // Create sample if not exists.
+    try {
+      sample.uuid ??= await dispatch(createSample(activeProjectUuid, name, sampleType));
```

Comment:
what does `??=` mean?

Reply:
it is to assign to uuid if it is not defined already, it was taken over form api v1

Reply:
yes, I noticed that we had it in v1 too, I wasn't sure about the syntax